Firestore/Example/Firestore_Benchmarks_iOS/FSTLevelDBBenchmarkTests.mm

```diff
@@ -0,0 +1,180 @@
+  NSString *dir =
+      [NSTemporaryDirectory() stringByAppendingPathComponent:@"FSTPersistenceTestHelpers"];
+  if ([files fileExistsAtPath:dir]) {
+    // Delete the directory first to ensure isolation between runs.
+    BOOL success = [files removeItemAtPath:dir error:&error];
+    if (!success) {
+      [NSException raise:NSInternalInconsistencyException
+                  format:@"Failed to clean up leveldb path %@: %@", dir, error];
+    }
+  }
+  return dir;
+}
+
+static FSTLevelDB *LevelDBPersistence() {
+  // This owns the DatabaseIds since we do not have FirestoreClient instance to own them.
+  static DatabaseId database_id{"p", "d"};
+
+  NSString *dir = LevelDBDir();
+  FSTSerializerBeta *remoteSerializer = [[FSTSerializerBeta alloc] initWithDatabaseID:&database_id];
+  FSTLocalSerializer *serializer =
+      [[FSTLocalSerializer alloc] initWithRemoteSerializer:remoteSerializer];
+  FSTLevelDB *db = [[FSTLevelDB alloc] initWithDirectory:dir serializer:serializer];
+  NSError *error;
+  BOOL success = [db start:&error];
+  if (!success) {
+    [NSException raise:NSInternalInconsistencyException
+                format:@"Failed to create leveldb path %@: %@", dir, error];
+  }
+
+  return db;
+}
+
+class LevelDBFixture : public benchmark::Fixture {
+  virtual void SetUp(benchmark::State &state) {
+    _db = LevelDBPersistence();
+    FillDB();
+  }
+
+  virtual void TearDown(benchmark::State &state) {
+    _db = nil;
+  }
+
+  void FillDB() {
+    LevelDbTransaction txn(_db.ptr, "benchmark");
+
+    for (int i = 0; i < _numDocuments; i++) {
+      FSTDocumentKey *docKey =
+          [FSTDocumentKey keyWithPathString:[NSString stringWithFormat:@"docs/doc_%i", i]];
+      std::string docKeyString = [FSTLevelDBRemoteDocumentKey keyWithDocumentKey:docKey];
+      txn.Put(docKeyString, DocumentData());
+      WriteIndex(txn, docKey);
+    }
+    txn.Commit();
+    // Force a write to disk to simulate startup situation
+    _db.ptr->CompactRange(NULL, NULL);
+  }
+
+ protected:
+  void WriteIndex(LevelDbTransaction &txn, FSTDocumentKey *docKey) {
+    txn.Put([FSTLevelDBDocumentTargetKey keyWithDocumentKey:docKey targetID:_targetID],
+            _emptyBuffer);
+    txn.Put([FSTLevelDBTargetDocumentKey keyWithTargetID:_targetID documentKey:docKey],
+            _emptyBuffer);
+  }
+
+  FSTLevelDB *_db;
+  // Arbitrary target ID
+  FSTTargetID _targetID = 1;
+  int _numDocuments = 10;
+  std::string _emptyBuffer;
+};
+
+// Plan: write a bunch of key/value pairs w/ empty strings (index entries)
+// Write a couple large values (documents)
+// In each test, either overwrite index entries and documents, or just documents
+
+BENCHMARK_DEFINE_F(LevelDBFixture, RemoteEvent)(benchmark::State &state) {
+  bool writeIndexes = (bool)state.range(0);
+  int64_t documentSize = state.range(1);
+  int64_t docsToUpdate = state.range(2);
+  std::string documentUpdate = UpdatedDocumentData((int)documentSize);
+  for (auto _ : state) {
```

Comment:
Can you make the loop variable a const reference? `for (const auto& _ : state) {`. While `_` variable is not used, copies are still being made. I don't know what `state` contains, but presumably these are objects larger than a couple of `int`s (otherwise, just leave as is).

Reply:
Done.